base/inference.jl

```diff
@@ -778,6 +793,253 @@ function limit_type_depth(t::ANY, d::Int, cov::Bool, vars::Vector{TypeVar}=TypeV
+    # detect cases where the comparison is trivial
+    if t === c
+        return false
+    elseif t === Union{}
+        return false # Bottom is as simple as they come
+    elseif isa(t, DataType) && isempty(t.parameters)
+        return false # fastpath: unparameterized types are always finite
+    elseif tupledepth > 0 && isa(unwrap_unionall(t), DataType) && isa(c, Type) && c !== Union{} && c <: t
+        return false # t is already wider than the comparison in the type lattice
+    elseif tupledepth > 0 && is_derived_type_from_any(unwrap_unionall(t), sources)
+        return false # t isn't something new
+    end
+    # peel off wrappers
+    if isa(c, UnionAll)
+        # allow wrapping type with fewer UnionAlls than comparison if in a covariant context
+        if !isa(t, UnionAll) && tupledepth == 0
+            return true
+        end
+        t = unwrap_unionall(t)
+        c = unwrap_unionall(c)
+    end
+    # rules for various comparison types
+    if isa(c, TypeVar)
+        if isa(t, TypeVar)
+            return !(t.lb === Union{} || t.lb === c.lb) || # simplify lb towards Union{}
+                   type_more_complex(t.ub, c.ub, sources, tupledepth)
+        end
+        c.lb === Union{} || return true
+        return type_more_complex(t, c.ub, sources, max(tupledepth, 1)) # allow replacing a TypeVar with a concrete value
+    elseif isa(c, Union)
+        if isa(t, Union)
+            return type_more_complex(t.a, c.a, sources, tupledepth) ||
+                   type_more_complex(t.b, c.b, sources, tupledepth)
+        end
+        return type_more_complex(t, c.a, sources, tupledepth) &&
+               type_more_complex(t, c.b, sources, tupledepth)
+    elseif isa(t, Int) && isa(c, Int)
+        return t !== 1 # alternatively, could use !(0 <= t < c)
+    end
+    # base case for data types
+    if isa(t, DataType)
+        tP = t.parameters
+        if isa(c, DataType) && t.name === c.name
+            cP = c.parameters
+            length(cP) < length(tP) && return true
+            ntail = length(cP) - length(tP) # assume parameters were dropped from the tuple head
+            # allow creating variation within a nested tuple, but only so deep
+            if t.name === Tuple.name && tupledepth > 0
+                tupledepth -= 1
+            elseif !isvarargtype(t)
+                tupledepth = 0
+            end
+            isgenerator = (t.name.name === :Generator && t.name.module === _topmod(t.name.module))
+            for i = 1:length(tP)
+                tPi = tP[i]
+                cPi = cP[i + ntail]
+                if isgenerator
+                    let tPi = unwrap_unionall(tPi),
+                        cPi = unwrap_unionall(cPi)
+                        if isa(tPi, DataType) && isa(cPi, DataType) &&
+                                !tPi.abstract && !cPi.abstract &&
+                                sym_isless(cPi.name.name, tPi.name.name)
```

Comment:
This looks like it could lead to really unpredictable effects.

I also think `type_more_complex` and `is_derived_type` could use a whole bunch of unit tests, mostly as examples illustrating what they do.